### ISSUE TITLE
Feature request: switch to work with student repositories named in GitHub Classroom style

### ISSUE THREAD

**Issue by DanielEWeeks** [repobee/repobee#694]
**Feature Request:**

I've been using GitHub Classroom, which automatically creates student repositories named in this style:  `<assignment-name>-<username>`.

It would be nice to have a command-line switch or setting that would make it easy to use student repositories named in the style `<assignment-name>-<username>` instead of the default `<username>-<assignment-name>` style.

Looks like I could change this by a simple edit to the `generate_repo_name` function in the `name.py` file to this line:

```
return "{}-{}".format(team_name, assignment_name)
```

Thank you.

**Comment by slarse** [comment]
Hi!

This is a fantastic idea, I'll definitely be making this possible in the near future. 

> Looks like I could change this by a simple edit to the generate_repo_name function in the name.py file to this line:

Yes, changing `generate_repo_name` to return `f"{assignment_name}-{team_name}"` should pretty much do it, but of course it's not a great long-term solution. I see two potential long-term solutions.

1. Add a CLI argument `--repo-name-format` that accepts some form of format string (e.g. something on the form `%A-%T` or `{assignment}-{team}`). Of course, the argument should also be configurable in the config file such that you don't have to repeat it all the time.
2. Make `generate_repo_name` a hook function, such that it can be implemented by a plugin.
3. Add both, and have `--repo-name-format` dynamically create a plugin that implements the `generate_repo_name` function.

Option 1 is the simpler option from a usability standpoint (I think?), but option 2 is more powerful and also easier to implement. Intuitively, option 3 seems enticing as it should be relatively easy to implement.

Any thoughts or preferences on this? I can probably get around to implementing a solution around mid-October.

**Comment by slarse** [comment]
Oh, right, there's one caveat that I didn't consider. RepoBee uses a team-based access model such that teams get access to repos, and students are added to the teams. Classroom uses direct repo access, where students are granted access directly to the repos.

Cloning repos and issue management with RepoBee will still be fine as that doesn't rely on the access model at all, it just clones based on the naming convention. Other features that do rely on the access model may not work. however. For example, the peer review functionality will most definitely not work, and the `--discover-repos` CLI option will also fail. Some plugins may also fail on the assumption that the student teams exist.

One could solve such compatibility issues by running the `repobee repos setup` command using dummy template repos, before starting to use RepoBee with Classroom-managed repos. That will set up all teams and give them access to the student repositories, but RepoBee won't try to push any files to repositories that already exists (i.e. the ones managed by Classroom). I don't think that should interfere with Classroom.

Was there any specific usage you had in mind?

**Comment by DanielEWeeks** [comment]
Thank you for considering this feature request.  Would be wonderful if RepoBee is extended to allow it to be used with GitHub Classroom repositories (Wish I had heard about RepoBee before I started using GitHub Classroom).

I was hoping to use RepoBee's peer review feature, so am sorry to hear that it isn't as simple as changing the style of naming of the student repositories to match that used by GitHub Classroom.  

Regarding implementation of an option to allow changing of the generate_repo_name function, option 3 sounds most flexible.  
 
Yes, as far as I understand, GitHub Classroom uses team-based access for group assignments, and direct access for individual assignments.

Thank you.

**Comment by slarse** [comment]
Alright, then I know what you're aiming for!

Off the top of my head, full interoperability with individual assignments in Classroom should be fairly simple. RepoBee's naming conventions effectively apply, just in the reverse order. It would be a matter of implementing option 3. as mentioned above, and then the user (e.g. you) would have to run a single command to create the teams that RepoBee expects to find. Then everything should work as expected. It could be made more intuitive by creating a Classroom plugin for RepoBee that adds a dedicated command for creating the expected teams. As Classroom is such a large part of VCS in education, I think a dedicated plugin to facilitate interoperability makes sense.

As for group assignments, I'm a bit uncertain. The name of the repo will be determined by the name of the team, and the students themselves effectively decide the names of the teams as there is no way to pre-allocate them. As such, no naming convention applies in the general case. I also can't find an API for Classroom, so I don't think there's a programmatic way to access wherever they store the information of which teams belong to which task. This information can of course be inferred relatively easily, but there is always the risk of unresolvable ambiguity.

So, to summarize, I think I can have a solution for interoperability with individual assignments sometime next month, but group assignments may prove a bit tricky.

Thanks again for the suggestion. This seems like such an obvious feature that I can't understand why it never even occurred to any of us working on and using RepoBee. I'll start working on this as soon as I'm back from vacation.

**Comment by DanielEWeeks** [comment]
Yes, the group assignments could be more tricky.  But for my current group assignments where the whole class is working on a single repository, I created a group named "Everyone", and when I create the group assignment in GitHub Classroom, I specify that is the group that will be used.  So for at least those, I know what the group name is, and it has been consistent across the group assignments so far.

For group assignments where the students choose the group name, couldn't RepoBee read a list of group names from an input file (similar to the 'students.txt' file)?  I guess that would work when the number of groups is small enough for the instructor to easily manually construct a list of the group names.

Thank you.  Hope you have a wonderful vacation.

**Comment by slarse** [comment]
> For group assignments where the students choose the group name, couldn't RepoBee read a list of group names from an input file (similar to the 'students.txt' file)?  I guess that would work when the number of groups is small enough for the instructor to easily manually construct a list of the group names.

Ah, yes, this could definitely be done rather easily. Generating such an input file could also be done semi-automatically, e.g. by RepoBee presenting the user with a list of teams it finds in the organization, and the user selecting which ones should be included in the list.

> Hope you have a wonderful vacation. 

Thanks!

**Comment by DanielEWeeks** [comment]
OK, as you suggested, after changing the generate_repo_name function to match GitHub Classroom's naming convention, and then doing `repos setup`, it appears that the student teams are set up correctly, and after that, all the peer review commands (assign, check, end) work perfectly in my initial (limited) testing.  Very nice!  Tomorrow I will try this 'peer review' feature with my class on their latest homework.   

Very minor comment:

This line in the ReproBee documentation:

```
$ repobee assign-reviews -a task-1 --sf students.txt --num-reviews 2
```
should now be

```
$ repobee reviews assign -a task-1 --sf students.txt --num-reviews 2
```

**Comment by slarse** [comment]
Cool!

Yup, that's a missed leftover from RepoBee 2.4, I've opened a separate issue on it. Thanks!

**Comment by DanielEWeeks** [comment]
The testing I did above used RepoBee-deployed test repositories that already had the team direct accesses auto-setup.  

Now if I try to do `reviews assign` on a set of previously-deployed GitHub Classroom repositories, it fails because the team direct write access has not been set:

```
$ repobee reviews assign -a 3-r-characters-homework --sf dw_students.txt --num-reviews 1 --issue review_issue.md
Fetching teams and repos: 100%| 2/2 [00:00<00:00,  2.28it/s]
[ERROR] NotFoundError: Can't find repos: 3-r-characters-homework-DanielEWeeks, 3-r-characters-homework-daniel-e-weeks
[ERROR] RepoBee exited unexpectedly. Please visit the FAQ to try to resolve the problem: https://repobee.readthedocs.io/en/stable/faq.html
```

The GItHub Classroom-created repositories `3-r-characters-homework-DanielEWeeks` and `3-r-characters-homework-daniel-e-weeks` exist, but the `DanielEWeeks` team and the `daniel-e-weeks` team did not have direct write access to their own homework repositories.

After adding the required team direct accesses, then the `review assign` command worked.

So if someone has already deployed a GitHub Classroom assignment, to use RepoBee with that set of assignments, one needs to do a `repos setup` type of command where it doesn't create the already existing repositories, but assigns the teams.

Is it save to apply 'repos setup' on an already deployed set of repositories that the students have already worked on?  That's not clear from the documentation - wouldn't want to inadvertently wipe out students' work in their repositories by mistake.

**Comment by DanielEWeeks** [comment]
Update using `RepoBee` with GitHub Classroom deployed assignments:

`repos setup` does not touch/alter already deployed sets of repositories, but does need to be run initially for the full list of students to set up the student teams.

After that, `review assign` worked well in assigning two peer reviewers to review their fellow student's repository, setting up the review teams and giving read access to the review team.  However, only opened a peer review issue for the first student in the list of two peer reviewers.

**Comment by slarse** [comment]
> However, it only opened a peer review issue for the first student in the list of two peer reviewers.

Just for the record, this problem is further discussed in #696 

Hi again @DanielEWeeks, hope you've gotten some use out of RepoBee in the past month! I'm back on the job now, so development on RepoBee will start to pick up again.

In any case, #713 adds a new hook function that allows plugins to override the naming scheme. I also intend to add a built-in plugin for GitHub Classroom interoperability, which will implement said hook, along with documentation on how to get started. I'd very much appreciate your feedback on that once I've added it.

I've for the time being decided against providing a CLI option to change the naming scheme. The reason is that I think that this feature won't be used much, and if you were to use it you'd probably want to use a single naming scheme permanently, as opposed to switching between different ones. Therefore, I think it's fully sufficient to allow changes to the name scheme through plugins (as they are so simple to implement, see #713 for an example).

**Comment by slarse** [comment]
@DanielEWeeks Docs for the [GitHub Classroom plugin are here](https://repobee.readthedocs.io/en/latest/builtins.html#ghclassroom), anything you think should be added at this point?

**Comment by DanielEWeeks** [comment]
Yes, after activating the GitHub Education 'unlimited teams' benefit, the peer review feature has been working seamlessly.  Very happy about that, as I think the peer review experience has been beneficial and educational for our students.  Thank you again for creating RepoBee and making peer review possible!

Happy also to see that you've now created the plugin for changing the repository naming scheme to allow easy usage of RepoBee with GitHub Classroom-created repositories.   

Regarding this "The first time you use RepoBee with GitHub Classroom-created repositories, you should run the `repos setup` command", does this need to be done only the first time or does it need to be done with each new assignment? 

I thought I needed to do `repos setup` with each new assignment.  I've been deploying new assignments via GitHub Classroom, and after all the students have accepted the assignment, generating their individual assignments, I then do `repos setup` before initiating peer review via the `reviews assign` command.

Thanks.

**Comment by slarse** [comment]
> I thought I needed to do `repos setup` with each new assignment

This is the correct. Poor phrasing on my part, will update the docs to reflect this. Thanks!

**Comment by slarse** [comment]
Oh, and FYI I plan to do a release of version 3.3 next Sunday (the 25th), and all of the recent updates will come along in that.

**Comment by slarse** [comment]
@DanielEWeeks [Docs updated](https://repobee.readthedocs.io/en/latest/builtins.html#ghclassroom), I think it should be more clear now! I'm also going to add this as a feature highlight somewhere as it's probably not very intuitive that this use case exists.

I'm closing this issue now as I think that it's been addressed. Thanks a lot for your input and feedback, it helps a lot in improving the product. Most users are silent users, unfortunately, so every little piece of feedback is at this point worth its digital weight in gold.

If you feel like RepoBee is helpful to you, consider starring the repo, and perhaps subtly mentioning it to your co-workers when they least expect it. I want the project to grow, but as it turns out my marketing skills leave some room for improvement :)
